src/pulsar_socket_writer.erl

```diff
@@ -0,0 +1,146 @@
+%% Copyright (c) 2025 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+
+-module(pulsar_socket_writer).
+
+-behaviour(gen_server).
+
+%% API
+-export([
+    start_link/4,
+    stop/1,
+
+    send_batch_async/7,
+    ping_async/1,
+    pong_async/1
+]).
+
+%% `gen_server' API
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2
+]).
+
+%%------------------------------------------------------------------------------
+%% Type declarations
+%%------------------------------------------------------------------------------
+
+%% Calls/Casts/Infos
+-record(send_batch, {topic, encoded_packet, num_messages}).
+-record(ping, {}).
+-record(pong, {}).
+
+%%------------------------------------------------------------------------------
+%% API
+%%------------------------------------------------------------------------------
+
+start_link(PartitionTopic, Host, Port, Opts) ->
+    case pulsar_socket:connect(Host, Port, Opts) of
+        {ok, Sock} ->
+            Params = #{partition_topic => PartitionTopic, sock => Sock, opts => Opts},
+            case gen_server:start_link(?MODULE, Params, []) of
+                {ok, SockPid} ->
+                    {ok, {SockPid, Sock}};
+                Error ->
+                    pulsar_socket:close(Sock, Opts),
+                    Error
+            end;
+        Error ->
+            Error
+    end.
+
+stop(SockPid) when is_pid(SockPid) ->
+    link(SockPid),
```

Comment:
this will cause caller to exit if caller is not trapping exit. but it has to happen, it's the remote-shell process (when trying to manually stop).